### ISSUE TITLE
`pub type Foo = foo::Foo` causes bad dead_code warnings on enum variants

### ISSUE THREAD

**Issue by kballard** [rust-lang/rust#17632]
The following code triggers `dead_code` warnings on each enum variant:

``` rust
pub type Foo = foo::Foo;
pub mod foo {
    pub enum Foo {
        One,
        Two
    }
}
```

This happens even though the enum is fully public. Commenting put the `pub type` line gets rid of the warnings.

**Comment by dherman** [comment]
I ran into a similar issue:
```js
mod tc39 {
    #[derive(Debug, Copy, Eq, PartialEq)]
    pub enum Politics {
        Good,
        Bad,
        Ugly
    }
}

use tc39::Politics::*;

fn main() {
    match Good {
        Good => println!("good!"),
        Bad => println!("bad!"),
        Ugly => println!("ugly!")
    }
}
```
This warns that `Bad` and `Ugly` aren't used.

**Comment by sanxiyn** [comment]
@kballard's code does not warn any more. @dherman's code does.

**Comment by birkenfeld** [comment]
@dherman's still does, but isn't that correct? The module is not public, neither is the import. If you make the `use tc39::Politics::*` a `pub use`, there is no warning anymore.

**Comment by sanxiyn** [comment]
This seems to be a case of #19140. Use in pattern doesn't count as use.

**Comment by Mark-Simulacrum** [comment]
Yes, we don't consider this sort of use a use, since it doesn't construct the value. Closing in favor of #19140.
